# Colorectal Cancer Meta-Analysis

TODO

## Requirements

- This project requires R version __3.5.1__ -- "Feather Spray"

- Please make sure that you have all required packages installed. You can
easily check whether that is the case by running

    ```bash
    Rscript requirements.R
    ```
    and confirm that it does not produce any errors.
- This project relies heavily on the R package `SIAMCAT`.  
    For the case that you are using an older version of the package, it will
    undoubtedly throw some errors. Thus, please make sure that you have the
    version __1.1.0__ installed. You can install the correct version of
    `SIAMCAT` via the
    [gitlab repository](https://git.embl.de/grp-zeller/SIAMCAT/tags/v1.1.0).
    Download the tar-ball of the package and then install the package by typing
    in R:

    ```R
    install.packages("SIAMCAT-v1.1.0.tar",
        repos = NULL, type="source")
    ```
    >- _you may have to change the path to the path of the downloaded file_
    >- _in Windows, you will need to have the package
    [rtools](https://cran.r-project.org/bin/windows/Rtools/) installed_
        >   - _i guess... never tested on a Windows system_

## Workflow

To reproduce the main figures of the project, you can follow these instructions.


### Setup

First, you will have to download the taxonomic and functional profiles and the
metadata needed for this project. They are stored on the ftp server of the
Zeller team and will be automatically downloaded and cleaned by running
this script.

```bash
cd ./src
Rscript prepare_data.R
```


### Association Testing

In order to run the association testing pipeline, please use the
`marker_analysis.R` script. The script computes the generalised fold change
(gFC), area under the Receiver-Operating-Characteristics curve (AUROC), and
the significance of single markers using a blocked Wilcoxon test. Additionally,
the script will run `ANCOM` for taxonomic profiles. The marker analysis can be
run for different sets of features, e.g. `KEGG` functional profiles and
species-level taxonomic profiles.

```bash
Rscript marker_analysis.R KEGG
Rscript marker_analysis.R species
```

The marker heatmap and the forest plot can be generated by calling this script:
```bash
Rscript figure_marker_heatmap.R species
```

>Note; this script may break if you use it for other features than species,
e.g. KEGG functional profiles   
in this case, you may need to manually tweak the script to get satisfying
results


### Confounder Analysis

To reproduce the confounder analysis and the accompanying figure, please type:

```bash
Rscript confounder_analysis.R species
```

The script will compute the variance explained by the disease status (i.e. CRC)
and by potential confounding variables (e.g. Age, Sex, BMI, etc.). The results
from this analysis motivated the inclusion of `Study` and `Colonoscopy` into
the differential abundance testing pipeline (by blocking for study and
colonoscopy status).  
The script will also produce a plot contrasting the variance explained by the
potential confounder and by disease status.


### Clustering

We found that the most strongly associated species can be grouped into four
different clusters with preferential enrichment in different patient subgroup.
The code to reproduce this analysis can be executed by running:
```bash
Rscript cluster_species.R
```
>Note: The ordering of the species in the heat map is not identical to the
ordering in the clustered histogram

The script will also produce a set of plots showing the positivity for each
species clusters in different patient subgroups.


### Machine Learning Pipeline

The machine learning pipeline can be run using the `train_models.R` script. The
script will train one LASSO model for each study and one model for each
leave-one-study-out setting. The models will be save in the respective
`./models/` folder. Please note that this step can take quite some time,
depending on the input features. For example, training all models for eggNOG
input features took approximately 8 hours on my machine.

After training the models, you can also reproduce the performance figure, using
the `figure_performance.R` script.

```bash
Rscript train_models.R species
Rscript figure_performance.R species
```
>Note: training all models for species input features took circa **45 min**
on my machine.

If you want to run the machine learning pipeline with another machine learning
method, you can
- either change the respective entry in the `parameters.yaml` file
- or you supply a second command line argument to `train_models.R`.

For example, if you want to train a Random Forest model and check the
performance, you can type:

```bash
Rscript train_models.R species randomForest
Rscript figure_performance.R species randomForest
```


### Functional Enrichment



### External Validation

There are two different types of external validations included in this project.

1. Other CRC studies  
To check the associations and the classification accuracy in other CRC studies,
you can run these scripts:
```bash
Rscript external_validation_associations.R
Rscript external_validation_classification.R species
```
Here, you can use all the models that have been trained before
(i.e. `species`/`KEGG`/`eggNOG`).  
Again, you can indicate another machine learning method as second command line
argument.

2. Non-CRC studies  
We also looked at other studies with shotgun metagenomic analyses that included
non-CRC patients to check how much cross-classification we get with other
diseases. In order to run this analysis, you can use:
```bash
Rscript external_validation_cross_classification.R
```
>Note: since this analysis only works on species-level taxonomic profiles as
input features, there is no argument to specify the features. But again,
there is an optional argument for the machine learning method to be used (if
the method differs from the method specified in the `parameters.yaml` file)


## Contact

If you have questions about the code or the analyses presented in this project,
please feel free to contact me, [Jakob Wirbel](mailto:jakob.wirbel@embl.de)
:smiley:
